keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,138 @@
+
+If the added parameter `forcePodRestart` is set to false, this UID should not be set, therefore keeping the podTemplate
+unchanged. This would prevent a rolling restart of all pods.
+
+Generally, the enhancer should not modify any template specs, but only the top-level resource if the flag is disabled.
+
+One option to consider is changing the behavior of the enhancer in general: At the moment, the enhancer adds labels and
+attributes to templates:
+```
+	fieldsToAdd := map[string]string{
+		kudo.HeritageLabel: "kudo",
+		kudo.OperatorLabel: metadata.OperatorName,
+		kudo.InstanceLabel: metadata.InstanceName,
+	}
+	labelPaths := [][]string{
+		{"metadata", "labels"},
+		{"spec", "template", "metadata", "labels"},
+		{"spec", "volumeClaimTemplates[]", "metadata", "labels"},
+		{"spec", "jobTemplate", "metadata", "labels"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "labels"},
+	}
+
+```
+The labels are very static no change is required here.
+
+
+```
+	fieldsToAdd := map[string]string{
+		kudo.PlanAnnotation:            metadata.PlanName,
+		kudo.PhaseAnnotation:           metadata.PhaseName,
+		kudo.StepAnnotation:            metadata.StepName,
+		kudo.OperatorVersionAnnotation: metadata.OperatorVersion,
+		kudo.PlanUIDAnnotation:         string(metadata.PlanUID),
+	}
+	annotationPaths := [][]string{
+		{"metadata", "annotations"},
+		{"spec", "template", "metadata", "annotations"},
+		{"spec", "jobTemplate", "metadata", "annotations"},
+		{"spec", "jobTemplate", "spec", "template", "metadata", "annotations"},
+	}
+```
+The annotations are volatile and will/may with every plan execution. The nested changes to template and jobTemplate
+are historically made because the behavior was modeled after Kustomize.
+
+This KEP proposes to change this: Attributes will only be applied to the top-level resource, not nested resources.
+
+The PlanUIDAnnotation *will* be set on nested resources and templates, depending on the new `forcePodReload` attribute.
+
+The values of the attributes can still be discovered by following the ownerReference of created resources.
+
+### Risks and Mitigations
+
+As we would skip the update of the template specs of pods, we would not write the latest plan, phase and step into
+the attributes of the pods. This could lead to incorrect data in the attributes, as the pod might still get restarted
+if the modified parameter is used in the template specs - it would be in the responsibility of the operator developer
+to prevent this scenario. 
+
+This new attribute has a very special meaning: It basically just prevents the update of attributes on template specs. It
+might be a too complex concept. This should be mitigated by careful naming of the new attribute and documentation.
+
+## Implementation History
+
+- 2020-03-31 - Initial draft. (@aneumann)
+
+## Alternatives
+
+A better alternative would be to have a full dependency graph of all resources to parameters. This would need to include
```

Comment:
I'm not sure I understand the alternative: even if we had a graph of all resources, a human operator would still want to decide whether to update parts of it or not, right?

Reply:
Well, a full dependency graph with all resources and parameters would solve the current problem for C*: If only node-count was changed, the pod template wouldn't need to be updated.
If a parameter changes that modifies a ConfigMap, the dependency graph would determine that this affects the pod, and generate a new UID and therefore restart the pod.

But yes, an operator might still need to define if Pods get restarted if a ConfigMap changes or not.

Reply:
Maybe worth putting this into the KEP ^^

Reply:
Done.